src/Iteration.cpp

```diff
@@ -447,6 +437,119 @@ Iteration::read()
+    // Initialize file with this to quiet warnings
+    // The following switch is comprehensive
+    Attributable * file = this;
+    switch( *series.m_iterationEncoding )
+    {
+        case IE::fileBased:
+            file = this;
+            break;
+        case IE::groupBased:
+            file = &series;
+            break;
+    }
+    AdvanceStatus status = series.advance(
+        AdvanceMode::BEGINSTEP, *file, myIteration< iterator_t >(), *this );
+    if( status != AdvanceStatus::OK )
+    {
+        return status;
+    }
+
+    // re-read -> new datasets might be available
+    if( *series.m_iterationEncoding == IE::groupBased &&
+        ( this->IOHandler->m_frontendAccess == Access::READ_ONLY ||
+          this->IOHandler->m_frontendAccess == Access::READ_WRITE ) )
+    {
+        bool previous = series.iterations.written();
+        series.iterations.written() = false;
+        auto oldType = this->IOHandler->m_frontendAccess;
+        auto newType =
+            const_cast< Access * >( &this->IOHandler->m_frontendAccess );
+        *newType = Access::READ_WRITE;
+        series.readGroupBased( false );
+        *newType = oldType;
+        series.iterations.written() = previous;
+    }
+
+    return status;
+}
+
+void
+Iteration::endStep()
+{
+    using IE = IterationEncoding;
+    Series & series =
+        *dynamic_cast< Series * >( parent->attributable->parent->attributable );
+    // Initialize file with this to quiet warnings
+    // The following switch is comprehensive
+    Attributable * file = this;
+    switch( *series.m_iterationEncoding )
+    {
+        case IE::fileBased:
+            file = this;
+            break;
+        case IE::groupBased:
+            file = &series;
+            break;
+    }
+    // @todo filebased check
+    series.advance(
+        AdvanceMode::ENDSTEP, *file, myIteration< iterator_t >(), *this );
+}
+
+template< typename Iterator >
+Iterator
+Iteration::myIteration()
+{
+    /* Find the root point [Series] of this file,
+        * meshesPath and particlesPath are stored there */
+    Writable *w = this->parent;
+    while( w->parent )
+        w = w->parent;
+
+    auto s = dynamic_cast< Series* >( w->attributable );
+    if( s == nullptr )
+        throw std::runtime_error("[Iteration::close] Series* is a nullptr");
+    for( auto it = s->iterations.begin(); it != s->iterations.end(); ++it )
+    {
+        if( it->second.m_writable.get() == this->m_writable.get() )
+        {
+            return it;
+        }
+    }
+    throw std::runtime_error(
+        "[Iteration::close] Iteration not found in Series." );
+}
+
+
+StepStatus *
+Iteration::stepStatus()
+{
+    Series * s =
+        dynamic_cast< Series * >( parent->attributable->parent->attributable );
+    StepStatus * flag = nullptr;;
+    switch( *s->m_iterationEncoding )
+    {
+        using IE = IterationEncoding;
+        case IE::fileBased:
+            flag = &*this->m_stepStatus;
```

Comment:
I knew there'd be a comment on this one :D
> Can we return StepStatus by value?

No, the intent of this is that some methods (`Iteration::close` as well as methods in `(Read|Write)Iterations`) need to modify the current `stepStatus`. Depending on the chosen iteration layout, `stepStatus` is a property of the whole `Series` or of each `Iteration`, so I'll need to decide which one to modify dynamically. So, similar to our `Attributable::written` method, this one must hand out a modifiable reference/pointer. Just that I have admittedly developed a strong dislike towards returning references, due to numerous accidental copying bugs I have had because of that...

Thinking about it, maybe a regular setter would do?

Reply:
> Thinking about it, maybe a regular setter would do?

A regular setter now does.